### PR TITLE
[InstCombine] combine mul(abs(x),abs(y)) to abs(mul(x,y))

### LINKED ISSUE #78076: multiplicativity of abs
Alive2 Proof: https://alive2.llvm.org/ce/z/XEDy0f
Original program: https://godbolt.org/z/e1YhxzWEh

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixes: https://github.com/llvm/llvm-project/issues/78076
Alive2 Proof: https://alive2.llvm.org/ce/z/XEDy0f

### CODE REVIEW

**Comment by nikic** [comment]
I think that https://github.com/llvm/llvm-project/pull/78110 already subsumes this.

**Review by nikic — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -330,6 +330,18 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
       return BinaryOperator::CreateMul(X, X);
   }
 
+  {
+    Value *X, *Y;
+    // abs(X) * abs(Y) -> abs(X * Y)
+    if (I.hasNoSignedWrap() &&
+        match(Op0, m_Intrinsic<Intrinsic::abs>(m_Value(X), m_AllOnes())) &&
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:337]: m_AllOnes -> m_One is more natural here.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 2]
```diff
@@ -330,6 +330,18 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+        match(Op1, m_Intrinsic<Intrinsic::abs>(m_Value(Y), m_AllOnes())))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:338]: You will also need `m_OneUse`. Please also add multi-use tests.

Comment (HerrCai0907) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp]: Done. Thanks!

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**